minijinja/src/value/object.rs

```diff
@@ -101,3 +88,230 @@ impl<T: Object> Object for std::sync::Arc<T> {
         T::call(self, state, args)
     }
 }
+
+/// A kind defines the object's behavior.
+///
+/// When a dynamic [`Object`] is implemented, it can be of one of the kinds
+/// here.  The default behavior will be a [`Basic`](Self::Basic) object which
+/// doesn't do much other than that it can be printed.  For an object to turn
+/// into a [struct](Self::Struct) or [sequence](Self::Seq) the necessary kind
+/// has to be returned with a pointer to itself.
+///
+/// Today object's can have the behavior of structs and sequences but this
+/// might expand in the future.  It does mean that not all types of values can
+/// be represented by objects.
+#[non_exhaustive]
+pub enum ObjectKind<'a> {
+    /// This object is a basic object.
+    ///
+    /// Such an object has no attributes but it might be callable and it
+    /// can be stringified.  When serialized it's serialized in it's
+    /// stringified form.
+    Basic,
+
+    /// This object is a sequence.
+    ///
+    /// Requires that the object implements [`SeqObject`].
+    Seq(&'a dyn SeqObject),
+
+    /// This object is a struct (map with string keys).
+    ///
+    /// Requires that the object implements [`StructObject`].
+    Struct(&'a dyn StructObject),
+}
+
+/// Views an [`Object`] as sequence of values.
+///
+/// # Example
+///
+/// The following example shows how to implement a dynamic object which
+/// represents a sequence of three items:
+///
+/// ```
+/// use std::fmt;
+/// use minijinja::value::{Value, Object, ObjectKind, SeqObject};
+///
+/// #[derive(Debug, Clone)]
+/// struct Point(f32, f32, f32);
+///
+/// impl fmt::Display for Point {
+///     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+///         write!(f, "({}, {}, {})", self.0, self.1, self.2)
+///     }
+/// }
+///
+/// impl Object for Point {
+///     fn kind(&self) -> ObjectKind<'_> {
+///         ObjectKind::Seq(self)
+///     }
+/// }
+///
+/// impl SeqObject for Point {
+///     fn get_item(&self, idx: usize) -> Option<Value> {
+///         match idx {
+///             0 => Some(Value::from(self.0)),
+///             1 => Some(Value::from(self.1)),
+///             2 => Some(Value::from(self.2)),
+///             _ => None,
+///         }
+///     }
+///
+///     fn seq_len(&self) -> usize {
+///         3
+///     }
+/// }
+///
+/// let value = Value::from_object(Point(1.0, 2.5, 3.0));
+/// ```
+pub trait SeqObject {
+    /// Looks up an item by index.
+    ///
+    /// Sequences should provide a value for all items in the range of `0..seq_len`
+    /// but the engine will assume that items within the range are `Undefined`
+    /// if `None` is returned.
+    fn get_item(&self, idx: usize) -> Option<Value>;
+
+    /// Returns the number of items in the sequence.
+    fn seq_len(&self) -> usize;
```

Comment:
I would personally prefer a name like `num_items()` for more coherence with `get_item()`.

Reply:
I don't like `num_X` much because it's not really a word. Maybe `item_count`?